demos/common/ota_demo_helpers/ota_demo_helpers.c

```diff
@@ -208,3 +239,137 @@ void SubscriptionManager_RemoveCallback( const char * pTopicFilter,
     }
 }
 /*-----------------------------------------------------------*/
+
+bool Agent_MessageSend( const AgentMessageContext_t * pxMsgCtx,
+                        const void * pData,
+                        uint32_t ulBlockTimeMs )
+{
+    BaseType_t xQueueStatus = pdFAIL;
+
+    if( ( pxMsgCtx != NULL ) && ( pData != NULL ) )
+    {
+        xQueueStatus = xQueueSendToBack( pxMsgCtx->queue, pData, pdMS_TO_TICKS( ulBlockTimeMs ) );
+    }
+
+    return ( xQueueStatus == pdPASS ) ? true : false;
+}
+
+/*-----------------------------------------------------------*/
+
+bool Agent_MessageReceive( const AgentMessageContext_t * pxMsgCtx,
+                           void * pBuffer,
+                           uint32_t ulBlockTimeMs )
+{
+    BaseType_t xQueueStatus = pdFAIL;
+
+    if( ( pxMsgCtx != NULL ) && ( pBuffer != NULL ) )
+    {
+        xQueueStatus = xQueueReceive( pxMsgCtx->queue, pBuffer, pdMS_TO_TICKS( ulBlockTimeMs ) );
+    }
+
+    return ( xQueueStatus == pdPASS ) ? true : false;
+}
+
+/*-----------------------------------------------------------*/
+
+static void initializePool()
+{
+    bool bSemAssigned = false;
+
+    taskENTER_CRITICAL();
+    {
+        if( ucInitStatus == SEMAPHORE_NOT_INITIALIZED )
+        {
+            bSemAssigned = true;
+            ucInitStatus = SEMAPHORE_INIT_PENDING;
+        }
+    }
+    taskEXIT_CRITICAL();
+
+    if( bSemAssigned )
+    {
+        memset( ( void * ) xCommandStructurePool, 0x00, sizeof( xCommandStructurePool ) );
+        xFreeCommandStructMutex = xSemaphoreCreateCounting( MQTT_COMMAND_CONTEXTS_POOL_SIZE, MQTT_COMMAND_CONTEXTS_POOL_SIZE );
+        ucInitStatus = SEMAPHORE_INITIALIZED;
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+Command_t * Agent_GetCommand( uint32_t ulBlockTimeMs )
```

Comment:
Would it make more sense to use the latest queue based implementation that's in the [Labs repo](https://github.com/FreeRTOS/Lab-Project-coreMQTT-Agent/blob/main/lib/FreeRTOS/mqtt-agent-interface/freertos_command_pool.c)? It did fix a race condition that resulted from using !NONE to indicate structs in use.

Reply:
Thanks refactored the agent interface and updated with the latest code as in the Labs repo.
https://github.com/aws/amazon-freertos/pull/3074/commits/b7ccb0bc15dd6f100decba126832437cea387557